lib/segment/src/index/field_index/full_text_index/posting_list.rs

```diff
@@ -37,23 +40,305 @@ impl PostingList {
+        } else {
+            data.len() - chunks[chunk_index].offset as usize
+        }
+    }
+
+    fn find_chunk(&self, doc_id: &PointOffsetType, start_chunk: Option<usize>) -> Option<usize> {
+        let start_chunk = start_chunk.unwrap_or(0);
+        match self.chunks[start_chunk..].binary_search_by(|chunk| chunk.initial.cmp(doc_id)) {
+            // doc_id is the initial value of the chunk with index idx
+            Ok(idx) => Some(start_chunk + idx),
+            // chunk idx has larger initial value than doc_id
+            // so we need the previous chunk
+            Err(idx) if idx > 0 => Some(start_chunk + idx - 1),
+            Err(_) => None,
+        }
+    }
+
+    fn is_in_postings_range(&self, val: PointOffsetType) -> bool {
+        !self.chunks.is_empty() && val >= self.chunks[0].initial && val <= self.last_doc_id
+    }
+
+    fn decompress_chunk(
+        &self,
+        bitpacker: &BitPackerImpl,
+        chunk_index: usize,
+        decompressed: &mut [PointOffsetType],
+    ) {
+        assert!(decompressed.len() == BitPackerImpl::BLOCK_LEN);
+        let chunk = &self.chunks[chunk_index];
+        let chunk_size = Self::get_chunk_size(&self.chunks, &self.data, chunk_index);
+        let chunk_bits = (chunk_size * 8) / BitPackerImpl::BLOCK_LEN;
+        bitpacker.decompress_sorted(
+            chunk.initial,
+            &self.data[chunk.offset as usize..chunk.offset as usize + chunk_size],
+            decompressed,
+            chunk_bits as u8,
+        );
+    }
+}
+
+// Help structure to check if set of sorted values are in the compressed posting list
+// This help structure reuse the decompressed chunk to avoid unnecessary decompression
+pub struct CompressedPostingVisitor<'a> {
+    bitpacker: BitPackerImpl,
+    postings: &'a CompressedPostingList,
+
+    // Data for the decompressed chunk.
+    decompressed_chunk: [PointOffsetType; BitPackerImpl::BLOCK_LEN],
+
+    // Index of the decompressed chunk.
+    // It is used to shorten the search range of chunk index for the next value.
+    decompressed_chunk_idx: Option<usize>,
+
+    // start index of the decompressed chunk.
+    // It is used to shorten the search range in decompressed chunk for the next value.
+    decompressed_chunk_start_index: usize,
+
+    // Check if the checked values are in the increasing order.
+    #[cfg(test)]
+    last_checked: Option<PointOffsetType>,
+}
+
+impl<'a> CompressedPostingVisitor<'a> {
+    pub fn new(postings: &'a CompressedPostingList) -> CompressedPostingVisitor<'a> {
+        CompressedPostingVisitor {
+            bitpacker: BitPackerImpl::new(),
+            postings,
+            decompressed_chunk: [0; BitPackerImpl::BLOCK_LEN],
+            decompressed_chunk_idx: None,
+            decompressed_chunk_start_index: 0,
+            #[cfg(test)]
+            last_checked: None,
+        }
+    }
+
+    pub fn contains(&mut self, val: &PointOffsetType) -> bool {
```

Comment:
Also docstring might help

Reply:
I agree, the `&mut` tripped me as well on a `contains` method :+1:

Reply:
This method is defined is separate help structure which reuse decompressed data to avoid unnecessary decompression. Help structure was commented. Added comments also to this function. And renamed into `contains_next `